lib/bolt/project_migrator/modules.rb

```diff
@@ -0,0 +1,222 @@
+# frozen_string_literal: true
+
+require 'bolt/project_migrator/base'
+
+module Bolt
+  class ProjectMigrator
+    class Modules < Base
+      def migrate(project, modulepath)
+        return true unless ENV['BOLT_MODULE_FEATURE'] && project.modules.nil?
+
+        @outputter.print_message "Migrating project modules\n\n"
+
+        config     = project.project_file
+        puppetfile = project.puppetfile
+        moduledir  = project.managed_moduledir
+
+        # Notify user to manually migrate modules if using non-default modulepath
+        if modulepath != project.modulepath
+          @outputter.print_migrate_step(
+            "Project has a non-default configured modulepath, unable to automatically "\
+            "migrate project modules. To migrate project modules manually, see "\
+            "http://pup.pt/bolt-modules"
+          )
+          true
+        # Migrate modules from Puppetfile
+        elsif File.exist?(puppetfile)
+          migrate_modules_from_puppetfile(config, puppetfile, moduledir, modulepath)
+        # Migrate modules to updated modulepath
+        else
+          consolidate_modules(modulepath)
+          update_project_config([], config)
+        end
+      end
+
+      # Migrates modules by reading a Puppetfile and prompting the user for
+      # which ones are direct dependencies for the project. Once the user has
+      # selected the direct dependencies, this will resolve the modules, write a
+      # new Puppetfile, install the modules, and then move any remaining modules
+      # to the new moduledir.
+      #
+      private def migrate_modules_from_puppetfile(config, puppetfile_path, moduledir, modulepath)
+        require 'bolt/puppetfile'
+        require 'bolt/puppetfile/installer'
+
+        # Migrating modules includes prompting. Do nothing if stdin is not a tty.
+        unless $stdin.tty?
+          @outputter.print_migrate_step(
+            "stdin is not a tty, unable to automatically migrate project modules."\
+            "To migrate project modules manually, see http://pup.pt/bolt-modules"
+          )
+          return false
+        end
+
+        begin
+          @outputter.print_migrate_step("Parsing Puppetfile at #{puppetfile_path}")
+          puppetfile = Bolt::Puppetfile.parse(puppetfile_path, skip_unsupported_modules: true)
```

Comment:
This just deletes local modules, which *definitely* doesn't seem like what we want. I think erroring here if we can't mirgate modules is the right thing to do.

Reply:
It should just move local modules into `modules`... Anything that isn't parsed _and_ selected as a managed module will just be moved. The only modules that are deleted are those that are parsed and selected as managed modules.

Reply:
Ah, ok, I see what  Idid - project migrate then puppetfile install. I'm not sure how common that will be - do we want to raise an error if you have the 'modules' key and try to puppetfile install

Reply:
I think this PR just needed a rebase. The work in #2189 disables `puppetfile install` if `modules` is configured.